File: BRaVe-Mobile-Backend/Helpers/NonceUtil.cs
namespace BRaVe_Mobile_Backend.Helpers
{
    using System;
    using System.Security.Cryptography;
    using System.Text;

    public static class NonceUtil
    {
        // Reuse one RNG instance
        private static readonly RandomNumberGenerator Rng = RandomNumberGenerator.Create();

        /// <summary>Generate a 32-byte random nonce and return base64url (no padding).</summary>
        public static string Generate() => Generate(32);

        /// <summary>Generate an N-byte random nonce and return base64url (no padding).</summary>
        public static string Generate(int length = 32)
        {
            const string chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
            var random = new Random();
            return new string(Enumerable.Repeat(chars, length)
                .Select(s => s[random.Next(s.Length)]).ToArray());
        }


        /// <summary>Generate raw random bytes.</summary>
        public static byte[] GenerateBytes(int numBytes)
        {
            if (numBytes <= 0) throw new ArgumentOutOfRangeException(nameof(numBytes));
            var buf = new byte[numBytes];
            Rng.GetBytes(buf);
            return buf;
        }

        /// <summary>base64url encode (no padding).</summary>
        public static string ToBase64Url(byte[] bytes)
        {
            if (bytes == null) throw new ArgumentNullException(nameof(bytes));
            return Convert.ToBase64String(bytes).TrimEnd('=').Replace('+', '-').Replace('/', '_');
        }

        /// <summary>base64url decode.</summary>
        public static byte[] FromBase64Url(string s)
        {
            if (string.IsNullOrWhiteSpace(s)) throw new ArgumentNullException(nameof(s));
            string b64 = s.Replace('-', '+').Replace('_', '/');
            switch (b64.Length % 4) { case 2: b64 += "=="; break; case 3: b64 += "="; break; }
            return Convert.FromBase64String(b64);
        }
    }

}

File: BRaVeMobile/app/src/main/java/intl/iom/bravemobile/helpers/NonceUtil.java
package intl.iom.bravemobile.helpers;

import android.util.Base64;
import android.util.Xml;

import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;

public final class NonceUtil {
    private static final SecureRandom RNG = new SecureRandom();

    public static String generate() {
        return generate(32);
    }
    /** Generate a N-byte random nonce, base64url (no padding). */
    public static String generate(int numBytes) {
        byte[] buf = new byte[numBytes];
        RNG.nextBytes(buf);
        return b64u(buf);
    }

    public static String b64u(byte[] b) {
        return Base64.encodeToString(b, Base64.NO_WRAP);
        //return Base64.encodeToString(b, Base64.URL_SAFE | Base64.NO_WRAP | Base64.NO_PADDING);
        //return new String(b, StandardCharsets.UTF_8);
    }
    public static byte[] decode(String s) {
        //return Base64.decode(s, Base64.URL_SAFE | Base64.NO_WRAP);
        return Base64.decode(s, Base64.NO_WRAP);
        //return s.getBytes(StandardCharsets.UTF_8);
    }
}
